core/psh/ping/ping.c

```diff
@@ -167,120 +167,170 @@ static int psh_ping_recv(int fd, uint8_t *data, int len, char *infostr, int info
-			if (*end != '\0' || ping_common.timeout <= 200) {
-				fprintf(stderr, "ping: Wrong timeout value!\n");
-				return -EINVAL;
-			}
-			break;
-		default:
-		case 'h':
-			psh_ping_help();
-			return 0;
+			case 'c':
+				ping_common.cnt = strtoul(optarg, &end, 10);
+				if (*end != '\0' || ping_common.cnt <= 0) {
+					fprintf(stderr, "ping: Wrong count value!\n");
+					return 1;
+				}
+				break;
+			case 't':
+				ping_common.ttl = strtoul(optarg, &end, 10);
+				if (*end != '\0' || ping_common.ttl <= 0) {
+					fprintf(stderr, "ping: Wrong ttl value!\n");
+					return 1;
+				}
+				break;
+			case 'i':
+				ping_common.interval = strtoul(optarg, &end, 10);
+				if (*end != '\0' || ping_common.interval < 0) {
+					fprintf(stderr, "ping: Wrong interval value!\n");
+					return 1;
+				}
+				break;
+			case 'W':
+				ping_common.timeout = strtoul(optarg, &end, 10);
+				if (*end != '\0' || ping_common.timeout <= 100) {
+					fprintf(stderr, "ping: Wrong timeout value!\n");
+					return 1;
+				}
+				break;
+			case 's':
+				ping_common.reqsz = strtoul(optarg, &end, 10) + sizeof(struct icmphdr);
+				ping_common.respsz = sizeof(struct iphdr) + ping_common.reqsz;
+				if (*end != '\0' || ping_common.reqsz > 2040) {
+					fprintf(stderr, "ping: Wrong payload len\n");
+					return 1;
+				}
+				break;
+			default:
+			case 'h':
+				ping_help();
+				return 0;
 		}
 	}
 
 	if (argc - optind != 1) {
 		fprintf(stderr, "ping: Expected address!\n");
-		return -EINVAL;
+		return 1;
 	}
 
 	if (inet_pton(ping_common.af, argv[optind], &ping_common.raddr.sin_addr) != 1) {
 		fprintf(stderr, "ping: Invalid IP address!\n");
-		return -EINVAL;
+		return 1;
 	}
 
-	if ((fd = psh_ping_socketConfig()) <= 0)
-		return -EIO;
+	if ((req = malloc(ping_common.reqsz)) == NULL) {
+		fprintf(stderr, "ping: Out of memory!\n");
+		return 1;
+	}
 
-	psh_ping_reqInit(req, sizeof(req));
+	if ((resp = malloc(ping_common.respsz)) == NULL) {
+		fprintf(stderr, "ping: Out of memory!\n");
+		free(req);
+		return 1;
+	}
 
-	while (ping_common.cnt-- && !psh_common.sigint) {
-		clock_gettime(CLOCK_MONOTONIC, &ts);
-		sent = ts.tv_sec * 1000000 + ts.tv_nsec / 1000;
-		if (psh_ping_send(fd, req, sizeof(req)) < 0)	{
-			close(fd);
-			fprintf(stderr, "ping: Fail to send a packet!\n");
-			return -EINVAL;
-		}
+	if ((fd = ping_sockconf()) <= 0) {
+		free(req);
+		free(resp);
+		return 2;
+	}
 
-		bzero(infostr, sizeof(infostr));
-		bzero(resp, sizeof(resp));
-		if (psh_ping_recv(fd, resp, sizeof(resp), infostr, sizeof(infostr)) > 0) {
-			clock_gettime(CLOCK_MONOTONIC, &ts);
-			elapsed = ts.tv_sec * 1000000 + ts.tv_nsec / 1000 - sent;
-			printf("%s time=%lu.%02lu ms\n", infostr, (unsigned long) (elapsed / 1000),
-			                                 (unsigned long) ((elapsed % 1000) / 10));
-		}
+	ping_reqinit(req, ping_common.reqsz);
+
+	while (ping_common.cnt-- && !psh_common.sigint) {
+		if (ping_echoreply(fd, req, resp) != 0)
```

Comment:
Can we reflect in `main()` exit code if the ping succeeded or not? (I consider this base functionality - to be useable in automatic testing and other scripts).

The original PING has semantics:
_If ping does not receive any reply packets at all it will exit with code 1. If a packet count and deadline are both specified, and fewer than count packets are received by the time the deadline has arrived, it will also exit with code 1. On other error it exits with code 2. Otherwise it exits with code 0. This makes it possible to use the exit code to see if a host is alive or not._

Reply:
Done. I return 1, when we fail on sending/receiving packets and 2 on other errors (like wrong arguments, enomem or fail to open socket).